Remotion/ObjectBinding/Web/UI/Controls/Validation/BocListValidationFailureRepository.cs

```diff
@@ -0,0 +1,315 @@
+  /// </summary>
+  public class BocListValidationFailureRepository: IBocListValidationFailureRepository
+  {
+    private class HandleableFailure
+    {
+      public BusinessObjectValidationFailure Failure { get; }
+
+      public bool Handled { get; private set; }
+
+      public HandleableFailure (BusinessObjectValidationFailure failure)
+      {
+        Failure = failure;
+        Handled = false;
+      }
+
+      public void MarkAsHandled ()
+      {
+        Handled = true;
+      }
+    }
+
+    private class RowFailureContainer
+    {
+      private List<HandleableFailure>? _rowFailures;
+      private Dictionary<IBocColumnDefinitionWithValidationSupport, List<HandleableFailure>>? _cellFailures;
+
+      public void AddRowFailures (IEnumerable<BusinessObjectValidationFailure> failures)
+      {
+        _rowFailures ??= new List<HandleableFailure>();
+        _rowFailures.AddRange(failures.Select(e => new HandleableFailure(e)));
+      }
+
+      public void AddCellFailures (IBocColumnDefinitionWithValidationSupport cellDefinition, IEnumerable<BusinessObjectValidationFailure> failures)
+      {
+        List<HandleableFailure> cellFailures;
+        _cellFailures ??= new Dictionary<IBocColumnDefinitionWithValidationSupport, List<HandleableFailure>>();
+        if (_cellFailures.TryGetValue(cellDefinition, out var existingFailureList))
+        {
+          cellFailures = existingFailureList;
+        }
+        else
+        {
+          cellFailures = new List<HandleableFailure>();
+          _cellFailures[cellDefinition] = cellFailures;
+        }
+
+        cellFailures.AddRange(failures.Select(e => new HandleableFailure(e)));
+      }
+
+      public IReadOnlyCollection<HandleableFailure> GetRowAndCellFailures ()
+      {
+        var failures = new List<HandleableFailure>();
+        failures.AddRange(GetRowFailures());
+        failures.AddRange(GetAllCellFailuresForRow());
+        return failures;
+      }
+
+      public IReadOnlyCollection<HandleableFailure> GetRowFailures ()
+      {
+        if (_rowFailures == null)
+          return Array.Empty<HandleableFailure>();
+
+        return _rowFailures.Where(e => !e.Handled).ToList();
+      }
+
+      private IReadOnlyCollection<HandleableFailure> GetAllCellFailuresForRow ()
+      {
+        if (_cellFailures == null)
+          return Array.Empty<HandleableFailure>();
+
+        return _cellFailures.SelectMany(c => c.Value.Where(e => !e.Handled)).ToList();
+      }
+
+      public IReadOnlyCollection<HandleableFailure> GetCellFailures (IBocColumnDefinitionWithValidationSupport cellDefinition)
+      {
+        if (_cellFailures == null)
+          return Array.Empty<HandleableFailure>();
+
+        if (!_cellFailures.TryGetValue(cellDefinition, out var handleableFailures))
+          return Array.Empty<HandleableFailure>();
+
+        return handleableFailures.Where(e => !e.Handled).ToList();
```

Comment:
we should move the handled-check from RowFailure to the call site. it will simplify the code and have the same type (i.e. the outer type in this case) always be responsible for the Handled check. Right now, it's mixed and that doesn't help readability. It will also mean that the naming on RowFailure does not need to include "Unhandled"